authentication/src/main/java/io/camunda/authentication/CamundaOAuthPrincipalService.java

```diff
@@ -74,58 +81,74 @@ public CamundaOAuthPrincipalService(
 
   public OAuthContext loadOAuthContext(final Map<String, Object> claims)
       throws OAuth2AuthenticationException {
+    final var authContextBuilder = new AuthenticationContextBuilder();
+    final var principals = oidcPrincipalLoader.load(claims);
+    final var username = principals.username();
+    final var clientId = principals.clientId();
+
+    final var principalIdentifiers = new HashMap<EntityType, Set<String>>();
+
+    if (username == null && clientId == null) {
+      throw new OAuth2AuthenticationException(
+          new OAuth2Error(OAuth2ErrorCodes.INVALID_CLIENT),
+          "Neither username claim (%s) nor clientId claim (%s) could be found in the claims. Please check your OIDC configuration."
+              .formatted(usernameClaim, clientIdClaim));
+    }
+    if (username != null) {
+      authContextBuilder.withUsername(username);
+      principalIdentifiers.put(USER, Set.of(username));
+    }
+
+    if (clientId != null) {
+      authContextBuilder.withClientId(clientId);
+      principalIdentifiers.put(CLIENT, Set.of(clientId));
+    }
+
     final var mappings = mappingServices.getMatchingMappings(claims);
     final Set<String> mappingIds =
         mappings.map(MappingEntity::mappingId).collect(Collectors.toSet());
     if (mappingIds.isEmpty()) {
       LOG.debug("No mappings found for these claims: {}", claims);
+    } else {
+      principalIdentifiers.put(MAPPING, mappingIds);
     }
 
     final Set<String> groups;
     if (StringUtils.hasText(groupsClaim)) {
       groups = new HashSet<>(oidcGroupsLoader.load(claims));
     } else {
+      // TODO: Get groups for username and clientId
       groups =
-          groupServices.getGroupsByMemberIds(mappingIds, EntityType.MAPPING).stream()
+          groupServices.getGroupsByMemberIds(mappingIds, MAPPING).stream()
               .map(GroupEntity::groupId)
               .collect(Collectors.toSet());
+      if (!groups.isEmpty()) {
+        principalIdentifiers.put(GROUP, groups);
+      }
     }
 
-    final var roles = roleServices.getRolesByMappingsAndGroups(mappingIds, groups);
+    final var roles =
+        roleServices.findAll(
+            RoleQuery.of(
+                roleQuery ->
+                    roleQuery.filter(
+                        roleFilter -> roleFilter.memberIdsByType(principalIdentifiers))));
     final var roleIds = roles.stream().map(RoleEntity::roleId).collect(Collectors.toSet());
 
+    // TODO: Get tenants for username and clientId
```

Comment:
🔧 Generally it's a good practice to add the issue number where you'll implement this to the TODO, and add a comment to this issue.